File: examples/objc/complex/Advanced.m
// Complex ObjC fixture: classes, categories, protocols, try/catch,
// branches, loops.
#import <Foundation/Foundation.h>

@interface UserRepository : NSObject
@property (nonatomic, strong) NSMutableDictionary *users;
- (NSString *)findUser:(NSNumber *)userId;
- (BOOL)loadAllUsers;
@end

@implementation UserRepository

- (instancetype)init {
    self = [super init];
    if (self) {
        _users = [NSMutableDictionary dictionary];
    }
    return self;
}

- (NSString *)findUser:(NSNumber *)userId {
    if ([self.users objectForKey:userId] != nil) {
        return [self.users objectForKey:userId];
    }
    return nil;
}

- (BOOL)loadAllUsers {
    @try {
        for (int i = 0; i < 10; i++) {
            NSString *key = [NSString stringWithFormat:@"user_%d", i];
            [self.users setObject:key forKey:@(i)];
        }
        return YES;
    } @catch (NSException *e) {
        NSLog(@"load failed: %@", e);
        return NO;
    }
}

@end

NSString *escapeSQL(NSString *input) {
    return [input stringByReplacingOccurrencesOfString:@"'" withString:@"''"];
}

void dispatchToken(NSString *token) {
    if ([token length] == 0) {
        return;
    }
    if ([token hasPrefix:@"admin_"]) {
        runAdmin(token);
    } else {
        runUser(token);
    }
}

void processBatch(NSArray *tokens) {
    for (NSString *token in tokens) {
        dispatchToken(token);
    }
}

void runAdmin(NSString *token) {
    system([[NSString stringWithFormat:@"admin-task %@", token] UTF8String]);
}

void runUser(NSString *token) {
    system([[NSString stringWithFormat:@"user-task %@", token] UTF8String]);
}

File: examples/erlang/complex/advanced.erl
%% Complex Erlang fixture: pattern matching, list comprehensions,
%% case, try/catch, multi-clause functions.
-module(advanced).
-export([find_user/2, load_all_users/0, process_batch/1]).

find_user(Users, Id) ->
    case maps:find(Id, Users) of
        {ok, User} -> User;
        error -> undefined
    end.

load_all_users() ->
    try
        Users = [{I, "user_" ++ integer_to_list(I)} || I <- lists:seq(0, 9)],
        maps:from_list(Users)
    catch
        _:E ->
            io:format("load failed: ~p~n", [E]),
            #{}
    end.

escape_sql(Input) ->
    re:replace(Input, "'", "''", [global, {return, list}]).

process_batch([]) ->
    ok;
process_batch([Token | Rest]) ->
    case Token of
        "" -> process_batch(Rest);
        "admin_" ++ _ ->
            run_admin(Token),
            process_batch(Rest);
        _ ->
            run_user(Token),
            process_batch(Rest)
    end.

run_admin(Token) ->
    os:cmd("admin-task " ++ Token).

run_user(Token) ->
    os:cmd("user-task " ++ Token).
